src/hardhat/contracts/Staking/FraxUnifiedFarm_ERC20.sol

```diff
@@ -372,25 +488,41 @@ contract FraxUnifiedFarm_ERC20 is FraxUnifiedFarmTemplate {
     // Two different withdrawLocked functions are needed because of delegateCall and msg.sender issues (important for migration)
     function withdrawLocked(bytes32 kek_id, address destination_address) nonReentrant external returns (uint256) {
         require(withdrawalsPaused == false, "Withdrawals paused");
-        return _withdrawLocked(msg.sender, destination_address, kek_id);
+
+        // Backwards compat - lookup the index of the stake item first.
+        uint256 arr_idx = getLockedStakeArrIndex(msg.sender, kek_id);
+        return _withdrawLocked(msg.sender, destination_address, kek_id, arr_idx);
+    }
+
+    // Overloaded to take the arr_idx such that it can be computed off-chain.
+    // Two different withdrawLocked functions are needed because of delegateCall and msg.sender issues (important for migration)
+    function withdrawLocked(bytes32 kek_id, uint256 arr_idx, address destination_address) nonReentrant external returns (uint256) {
+        require(withdrawalsPaused == false, "Withdrawals paused");
+        return _withdrawLocked(msg.sender, destination_address, kek_id, arr_idx);
     }
 
     // No withdrawer == msg.sender check needed since this is only internally callable and the checks are done in the wrapper
     // functions like migrator_withdraw_locked() and withdrawLocked()
     function _withdrawLocked(
         address staker_address,
         address destination_address,
-        bytes32 kek_id
+        bytes32 kek_id,
+        uint256 arr_idx
     ) internal returns (uint256) {
         // Collect rewards first and then update the balances
         _getReward(staker_address, destination_address, true);
 
         // Get the stake and its index
-        (LockedStake memory thisStake, uint256 theArrayIndex) = _getStake(staker_address, kek_id);
+        LockedStake memory thisStake = _getStake(staker_address, kek_id, arr_idx);
         require(block.timestamp >= thisStake.ending_timestamp || stakesUnlocked == true || valid_migrators[msg.sender] == true, "Stake is still locked!");
         uint256 liquidity = thisStake.liquidity;
 
         if (liquidity > 0) {
+            // Should also add a check that _total_liquidity_locked|_locked_liquidity doesn't go negative?
+            // Perhaps overkill if no other bugs.
+            require(_total_liquidity_locked - liquidity > 0, "Withdrawing more liquidity than locked");
+            require(_locked_liquidity[staker_address] - liquidity > 0, "Withdrawing more liquidity than locked");
```

Comment:
also an overkill. if `lockAdditional` and `stakeLocked` lock the right amounts, the right liquidity should always be returned.